### ISSUE TITLE
~ appears as < for Logitech MX Keys

### ISSUE THREAD

**Issue by neofight78** [elementary/settings-keyboard#298]
I have a Russian edition Logitech MX Keys keyboard and I am using the Russian and English (US) layouts.

The English (US) layout seems to work fine except the ` & ~ key appears as < & >. I googled the following fix:

setxkbmap -option apple:badmap

Which seems to sort it out, but it seems the change doesn't always survive layout switches, reboots etc.

In any case this should be a problem I can sort out via the keyboard settings UI.

**Comment by neofight** [comment]
@davidmhewitt Any chance you could take a look at this as this month's sponsorship item?

**Comment by davidmhewitt** [comment]
Thanks for the sponsorship. I think I'll struggle to help because it is a hardware specific issue, but I have some questions.

1) Have you ever used this keyboard on Mac hardware or on MacOS? 
2) Have you tried the alternative layouts in the keyboard settings UI? Like "English (US) -> English (Macintosh)"?

**Comment by neofight** [comment]
Understood, I appreciate the help however far you get.

1. Yes, it works fine on a Mac. It wasn't automatically recognised, but the wizard popped up and I was asked to press the key immediately to the right of the left shift key. From that macOS magically worked out what they layout was.

2. Yes, I tried a whole bunch of layouts and English (US) seems the closest. 

Does the command I mentioned give any clues? Could it be added to the UI somehow? Alternatively a way to remap single keys might be ok?

**Comment by neofight** [comment]
Found this: https://askubuntu.com/questions/843590/key-keycode-49-and-key-keycode-94-are-inverted-on-macbookpro-keyboar/844931#844931

Is it something that could be exposed via the settings UI?

**Comment by davidmhewitt** [comment]
I think it's quite unlikely that any new UI to fix driver issues with keyboard is going to be accepted.

However, I can help you investigate more and maybe we'll find a more permanent fix. We might find that there's some config that we can submit to a hardware database that fixes this on all linux distros.

You'll need to install `libinput-tools` if you haven't already. You can do this with `sudo apt install libinput-tools`

Then if you run `sudo libinput list-devices`, you should get a list of the input devices attached to the system.

Find your keyboard in this list and take note of its `Kernel:` line. It will be something like `/dev/input/event0`.

Please show me the output of `udevadm info /dev/input/event0` (replace event0 with your value from the previous command).

Now install `evtest` (`sudo apt install evtest`)

Run it with `sudo evtest /dev/input/event0` (again, replace event0)

Press the keys you are having problems with (`~` and `<` ? ) and use `CTRL+C` to exit. Please copy and paste the output from this too so I can see.

**Comment by neofight** [comment]
Ok, this is run without the above fix I mentioned applied:

`udevadm info /dev/input/event12

P: /devices/virtual/misc/uhid/0005:046D:B35B.0002/input/input12/event12
N: input/event12
E: BACKSPACE=guess
E: DEVNAME=/dev/input/event12
E: DEVPATH=/devices/virtual/misc/uhid/0005:046D:B35B.0002/input/input12/event12
E: ID_BUS=bluetooth
E: ID_INPUT=1
E: ID_INPUT_KEY=1
E: ID_INPUT_KEYBOARD=1
E: LIBINPUT_DEVICE_GROUP=5/46d/b35b:04:33:C2:6B:D8:F6
E: MAJOR=13
E: MINOR=76
E: SUBSYSTEM=input
E: TAGS=:power-switch:
E: USEC_INITIALIZED=364764154
E: XKBLAYOUT=us,ru
E: XKBVARIANT=,
`
There is just one key I am having problems with. This is what I get for what should be the back tick (or tilde when holding shift)

`<Event: time 1596264291.238046, type 4 (EV_MSC), code 4 (MSC_SCAN), value 70064
Event: time 1596264291.238046, type 1 (EV_KEY), code 86 (KEY_102ND), value 0
`

**Comment by davidmhewitt** [comment]
Hey thanks, that's really interesting and helpful.

The backtick key you're pressing is in the top left of the keyboard, above tab?

It looks like your Logitech keyboard is pretending to be an Apple keyboard. I found a post on Reddit that I can't find the link to now that said that if it's been connected to a Mac running the Logitech config software, then the keyboard firmware switches into some kind of "Mac mode". Equally, if you connect it to a Windows PC running the Logitech software, it switches back. Do you have a Windows PC or partition you could try this with and see if it makes any difference when you next use it on elementary?

Or, do you have an alternative way to connect it to elementary OS that isn't bluetooth? Does it have the ability to be wired or connect via a USB dongle that isn't bluetooth?

I think we can probably tell the kernel to remap these keys, but I don't want to submit these fixes to the kernel or systemd if they're potentially going to mess up MX Keys keyboards in "Windows mode".

**Comment by neofight** [comment]
No, it's to the right of the left shift key (see below). It's an interesting theory about the modes, but it doesn't quite match what I am seeing 100%. You'll see next to the space bar there are two keys that when connected to my Mac are Opt & Cmd, but when connected to elementary OS (on PC hardware) are Start & Alt. i.e. the position of the super key changes. So presumably the mode is already being switched?

I'll try with the usb dongle shortly, a Windows machine will take a bit longer though. The first thing I did with my ThinkCentre Tiny was to overwrite Windows with elementary OS :)
 
![CleanShot 2020-08-01 at 21 01 53](https://user-images.githubusercontent.com/221644/89103218-55bf1800-d43a-11ea-9971-74419bb26680.png)

**Comment by neofight** [comment]
Ok, so connecting via the USB dongle instead of bluetooth makes no difference. I'll try testing with Windows a bit later.

**Comment by davidmhewitt** [comment]
Could you find the `/dev/input/eventX` path of your keyboard again (it sometimes changes) and run the following:

`cat /sys/class/input/eventX/device/modalias` (replace eventX)

I'll prepare a config file that will remap that key at a hardware level that you can try.

**Comment by neofight** [comment]
Here's it is:
```
input:b0005v046DpB35Be0011-e0,1,4,11,14,k71,72,73,74,75,77,79,7A,7B,7C,7D,7E,7F,80,81,82,83,84,85,86,87,88,89,8A,B7,B8,B9,BA,BB,BC,BD,BE,BF,C0,C1,C2,F0,ram4,l0,1,2,3,4,sfw
```

**Comment by davidmhewitt** [comment]
Ok,  create `/etc/udev/hwdb.d/10-mx-keys.hwdb` and put the following content:

```
evdev:input:b0005v046DpB35B*
  KEYBOARD_KEY_70064=grave
```

I'm not sure if the directory exists by default, so you may need to do this first:
`sudo mkdir -p /etc/udev/hwdb.d`

If you're not familiar with writing files into folders you need administrator permission for, you can do this with nano like this:

`sudo nano /etc/udev/hwdb.d/10-mx-keys.hwdb`

Then paste the contents from above and save and quit with `CTRL+O` then `CTRL+X`.

Once this is done, reboot and see if it helps.

**Comment by neofight** [comment]
Yes, that seems to solve it, thank you! Is this something that can be submitted somewhere to help (hopefully) more people than just me?

**Comment by davidmhewitt** [comment]
If you can try out the theory of the keyboard behaving differently after it's been used on Windows with the Logitech software. Run `evtest` commands again from above and see if there is a different scancode when you press that key.

If there's no difference to the `70064` scancode, then I'll submit this to the necessary place so that this can be used by all Linux distributions in the future. 

This config file should work for you on other Linux distributions too should you decide to switch away from elementary in the future.

Thanks again for the sponsorship!

**Comment by neofight** [comment]
Great, thanks for your help. It's very much appreciated. I'll test out the Windows theory as soon as I get a chance.

**Comment by lenemter** [comment]
Closing as stale
